### PR TITLE
feat: Design document for performance benchmarking

### PR DESCRIPTION
**Pull request by brian-intel**
## PR Checklist

- [x] Added label to the Pull Request for easier discoverability and search
- [x] Commit Message meets guidelines as indicated in the URL https://github.com/intel-retail/automated-self-checkout/blob/main/CONTRIBUTING.md
- [ ] PR introduces changes that breaks compatibility with other modules (If YES, please provide details below)
- [x] No Employer Confidential details are added

## What requirement is this design document for?

#3 

## If there are any other design Pull Requests or requirements, please link them here (i.e. intel-retail/automated-self-checkout )

None

### CODE REVIEW

File: Architecture/performance_benchmarking.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+To assist customers we will provide a set of performance Docker containers to measure the performance of their pipelines. The performance Docker containers will need to be supported on most modern Intel HW. The output will also need to be formatted and presented to customers as a hardware recommendation.
+
+## Proposed Design 
+
+<!-- Please provide a high level design of the proposed requirement. -->
+
+### Benchmark Script
+
+The benchmark script is designed to help determine the performance needs for a specific pipeline profile. The
```
Comment (jim-wang-intel) [inline Architecture/performance_benchmarking.md:35]: this paragraph is incomplete

File: Architecture/performance_benchmarking.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+sysstat: System CPU utilization
+free: System memory usage
+iotop: System Disk read and write data
+igt-gpu-tools: Integrated GPU utilization
+Intel XPU Manage: Discrete GPU utilization
+Intel Performance Counter Monitor: System power usage
+
+### Input Source Types
+
+For performance inputs we support RTSP, USB camera, Intel® RealSense™ Camera, and video files. For test consistency it is recommended to use a video loop with an RTSP stream. As an option a RTSP [Camera Simulator](https://intel-retail.github.io/automated-self-checkout/run_camera_simulator.html) is provided with the performance script.
```
Comment (jim-wang-intel) [inline Architecture/performance_benchmarking.md:47]:
```suggestion
For performance inputs we support RTSP video streaming, USB camera, Intel® RealSense™ Camera, and video files. For testing consistency it is recommended to use a video loop with an RTSP stream. As an option an RTSP [Camera Simulator](https://intel-retail.github.io/automated-self-checkout/run_camera_simulator.html) is provided with the performance script.
```

File: Architecture/performance_benchmarking.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+If you are looking to test a specific number of pipelines on different hardware SKUs the `--pipelines` parameter can be used. This parameter will start the specified number of pipelines 
+
+[Specified Number of Pipelines](https://intel-retail.github.io/automated-self-checkout/pipelinebenchmarking.html#benchmark-specified-number-of-pipelines)
+
+[![Specified Number of Pipelines](./images/num-of-pipelines.jpg)](./images/num-of-pipelines.jpg)
+
+#### Stream Density
+
+If looking for the maximum number of pipelines that are supported on a specific hardware SKU you can use the stream density parameter. By setting the `--stream_density` parameter to the desired frames per second (FPS) the script will continue to create pipelines until the average pipelines FPS falls below the desired FPS. The scrip will provide a detailed log to show each pipeline FPS during the test run. This option provides a method for testing the top performance when introducing a new pipeline or hardware SKU.
```
Comment (jim-wang-intel) [inline Architecture/performance_benchmarking.md:62]:
```suggestion
If looking for the maximum number of pipelines that are supported on a specific hardware SKU for a given target FPS,  you can use the stream density parameter. By setting the `--stream_density` parameter to the desired frames per second (FPS) the script will continue to create pipelines until the average pipelines FPS falls below the desired FPS. The scrip will provide a detailed log to show each pipeline FPS during the test run. This option provides a method for testing the top performance when introducing a new pipeline or hardware SKU.
```

**Review by jim-wang-intel — CHANGES_REQUESTED** [review]
some grammar fix and suggestions

File: Architecture/performance_benchmarking.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+sysstat: System CPU utilization
+free: System memory usage
+iotop: System Disk read and write data
+igt-gpu-tools: Integrated GPU utilization
+Intel XPU Manage: Discrete GPU utilization
+Intel Performance Counter Monitor: System power usage
+
+### Input Source Types
+
+For performance inputs we support RTSP video streams, USB camera, Intel® RealSense™ Camera, and video files. For longer benchmarking runs it's recommended to use a video loop with an RTSP stream for inference result consistency. As an option a RTSP [Camera Simulator](https://intel-retail.github.io/automated-self-checkout/run_camera_simulator.html) is provided with the performance script.
```
Comment (jim-wang-intel) [inline Architecture/performance_benchmarking.md:42]: sorry, i missed this one: 
```suggestion
For performance inputs we support RTSP video streams, USB camera, Intel® RealSense™ Camera, and video files. For longer benchmarking runs it's recommended to use a video loop with an RTSP stream for inference result consistency. As an option an RTSP [Camera Simulator](https://intel-retail.github.io/automated-self-checkout/run_camera_simulator.html) is provided with the performance script.
```

**Review by jim-wang-intel — APPROVED** [review]

File: Architecture/performance_benchmarking.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+## Decision
+
+<!-- Requirements approval board will update this section with justification for approval or rejection -->
+
+## Context  
+
+<!-- Please provide context to the requirement. -->
+
+To assist customers we will provide a set of performance Docker containers to measure the performance of their pipelines. The performance Docker containers will need to be supported on most modern Intel HW. The output will also need to be formatted and presented to customers as a hardware recommendation.
```
Comment (NeethuES-intel) [inline Architecture/performance_benchmarking.md:22]:
```suggestion
To assist customers we will provide a set of performance Docker containers to measure the performance of their pipelines. The performance Docker containers will need to be supported on most modern Intel hardware. The output will also need to be formatted and presented to customers as a hardware recommendation.
```

File: Architecture/performance_benchmarking.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+sysstat: System CPU utilization
+free: System memory usage
+iotop: System Disk read and write data
+igt-gpu-tools: Integrated GPU utilization
+Intel XPU Manage: Discrete GPU utilization
+Intel Performance Counter Monitor: System power usage
+
+### Input Source Types
+
+For performance inputs we support RTSP video streams, USB camera, Intel® RealSense™ Camera, and video files. For longer benchmarking runs it's recommended to use a video loop with an RTSP stream for inference result consistency. As an option an RTSP [Camera Simulator](https://intel-retail.github.io/automated-self-checkout/run_camera_simulator.html) is provided with the performance script.
```
Comment (NeethuES-intel) [inline Architecture/performance_benchmarking.md:42]:
```suggestion
For performance inputs we support RTSP video streams, USB camera, Intel® RealSense™ Camera, and video files. For longer benchmarking runs its' recommended to use a video loop with an RTSP stream for inference result consistency. As an option an RTSP [Camera Simulator](https://intel-retail.github.io/automated-self-checkout/run_camera_simulator.html) is provided with the performance script.
```

File: Architecture/performance_benchmarking.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+If you are looking to test a specific number of pipelines on different hardware SKUs the `--pipelines` parameter can be used. This parameter will start the specified number of pipelines 
+
+[Specified Number of Pipelines](https://intel-retail.github.io/automated-self-checkout/pipelinebenchmarking.html#benchmark-specified-number-of-pipelines)
+
+[![Specified Number of Pipelines](./images/num-of-pipelines.jpg)](./images/num-of-pipelines.jpg)
+
+#### Consolidated Results
+
+To make reading results easier a consolidation script has been provided. This script will work with a single or multiple runs of the specified number of pipelines. Details about this process are found in [Benchmark Specified Number of Pipelines](https://intel-retail.github.io/automated-self-checkout/pipelinebenchmarking.html#benchmark-specified-number-of-pipelines)
```
Comment (NeethuES-intel) [inline Architecture/performance_benchmarking.md:57]:
```suggestion
To make reading results easier, a consolidation script has been provided. This script will work with a single or multiple runs of the specified number of pipelines. Details about this process are found in [Benchmark Specified Number of Pipelines](https://intel-retail.github.io/automated-self-checkout/pipelinebenchmarking.html#benchmark-specified-number-of-pipelines)
```

File: Architecture/performance_benchmarking.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+``` bash
+make consolidate ROOT_DIRECTORY=<output dir>
+```
+
+#### Stream Density
+
+The stream density parameter can be used to find the maximum number of pipelines at a target frames per second (FPS) on a specific hardware SKU. By setting the `--stream_density` parameter to the desired FPS the script will continue to create pipelines until the average pipelines FPS falls below the desired FPS. The scrip will provide a detailed log to show each pipeline FPS during the test run. This option provides a method for testing the top performance when introducing a new pipeline or hardware SKU.
```
Comment (NeethuES-intel) [inline Architecture/performance_benchmarking.md:65]:
```suggestion
The stream density parameter can be used to find the maximum number of pipelines at a target frames per second (FPS) on a specific hardware SKU. By setting the `--stream_density` parameter to the desired FPS the script will continue to create pipelines until the average pipelines FPS falls below the desired FPS. The script will provide a detailed log to show each pipeline FPS during the test run. This option provides a method for testing the top performance when introducing a new pipeline or hardware SKU.
```

**Review by NeethuES-intel — APPROVED** [review]

**Pull request merged**